### ISSUE TITLE
checkbox example code bug?

### ISSUE THREAD

**Issue by DerekZH** [bokeh/bokeh#5743]
To learn to use Checkbox using CustomJS, I am trying to reproduce the plot in the code below:

https://github.com/bokeh/bokeh/blame/214442205411931c57da8b896bdf68a457eb8c66/examples/plotting/file/line_on_off.py

For some reason no matter which checkbox is clicked, only Line 2 is toggled on/off.  I observed the same behavior when I modified the code in my own plot.  

My bokeh version is:
bokeh==0.12.4

@birdsarah would appreciate if you might take a look.

**Comment by rocheseb** [comment]
The callback in the example is written in coffeescript and you would need to use CustomJS.from_coffeescript() instead of CustomJS() (and you will need to install nodejs)

If you want to use plain javascript with CustomJS you will need something like this:

```
var indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };

p0.visible = indexOf.call(checkbox.active,0)>=0;
```

**Comment by bryevdv** [comment]
Can everyone here please first confirm that they are running the version of the example that matches the version of bokeh they have installed?

**Comment by rocheseb** [comment]
I have bokeh 0.12.4. In the following code there is "codeold", which is the same as in the example and doesn't work (it always toggles lines in a specific order no matter which checkboxes you check) and "code" which works, this is with standalone html.

```
import numpy as np
import os

from bokeh.palettes import Viridis3
from bokeh.plotting import Figure
from bokeh.models import CheckboxGroup, CustomJS
from bokeh.layouts import gridplot
from bokeh.resources import CDN
from bokeh.embed import file_html

codeold = """
    if (0 in checkbox.active) {
        l0.visible = true
    } else {
        l0.visible = false
    }
    if (1 in checkbox.active) {
        l1.visible = true
    } else {
        l1.visible = false
    }
    if (2 in checkbox.active) {
        l2.visible = true
    } else {
        l2.visible = false
    }
"""

code = """
var indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };

l0.visible = indexOf.call(checkbox.active,0)>=0;
l1.visible = indexOf.call(checkbox.active,1)>=0;
l2.visible = indexOf.call(checkbox.active,2)>=0;
"""

p = Figure()
props = dict(line_width=4, line_alpha=0.7)
x = np.linspace(0, 4 * np.pi, 100)
l0 = p.line(x, np.sin(x), color=Viridis3[0], legend="Line 0", **props)
l1 = p.line(x, 4 * np.cos(x), color=Viridis3[1], legend="Line 1", **props)
l2 = p.line(x, np.tan(x), color=Viridis3[2], legend="Line 2", **props)

callback = CustomJS(code=code, args={})
checkbox = CheckboxGroup(labels=["Line 0", "Line 1", "Line 2"], active=[0, 1, 2], callback=callback, width=100)
callback.args = dict(l0=l0, l1=l1, l2=l2, checkbox=checkbox)

grid = gridplot([[checkbox,p]])
outfile=open(os.path.join('boktest.html'),'w')
outfile.write(file_html(grid,CDN,'test'))
outfile.close()
```

**Comment by mattpap** [comment]
The question is about an old and utterly wrong version of `plotting/file/line_on_off.py` example. The updated version uses coffeescript where `in` operator has decent meaning in array context.  `0 in checkbox.active` in JavaScript asks if object `checkbox.active` has property named `0`. This doesn't ask if `0` is an element of that array. Given how arrays are implemented in JavaScript, this also asks whether array item at index `0` was "touched" (assigned a value). To check if `0` is an element of an array, use `indexOf()` method (as suggested earlier). You can also skip the polyfill for `indexOf()`, unless you want to support IE <= 8, which we don't support anyway.

**Comment by bryevdv** [comment]
Right, I can't stress this enough: While there are some parts of Bokeh that are not very stable, and have been for years, Bokeh is a very large library and some parts are *still evolving, sometimes rapidly*. For instance include: `bokeh.charts`, custom extensions, etc. For this reason, ***it is imperative to always run the example that correspond exactly to your installed bokeh version***. I've tested the `0.12.4` version of the example:

https://github.com/bokeh/bokeh/blob/0.12.4/examples/plotting/file/line_on_off.py

(note the `0.12.4` in the URL) with an installed Bokeh version `0.12.4` and it worked as expected. 

Accordingly, I am closing this issue. Also as a gentle reminder, please start by taking general support requests (i.e., if something is not working but not 100% sure if its a library or user error) to the mailing list first, before the issue tracker.

**Comment by DerekZH** [comment]
Thank you guys for looking into this and the detailed explanation.  I followed the correct version of the code in my own plot, it works like a charm.  The capability of Bokeh is very very impressive!

I definitely was trying to run the example included in an older version using a newer version of Bokeh.  Both checking bokeh version and start the conversation in mailing list points well taken.  Thanks @bryevdv !

**Comment by bryevdv** [comment]
Glad it's working! closing now
